phase1/openstack/do

```diff
@@ -0,0 +1,103 @@
+#!/bin/bash
+
+set -o errexit
+set -o pipefail
+set -o nounset
+set -x
+
+cd "${BASH_SOURCE%/*}"
+
+CONFIG_FILE="${CONFIG_JSON_FILE:-../../.config.json}"
+CLUSTER_NAME=$(jq -r '.phase1.cluster_name' ${CONFIG_FILE})
+KUBE_CONTEXT_NAME=$(jq -r '.phase2.kube_context_name' ${CONFIG_FILE})
+TMP_DIR=${CLUSTER_NAME}/.tmp
+
+generate_token() {
+  perl -e 'printf "%06x.%08x%08x\n", rand(0xffffff), rand(0xffffffff), rand(0xffffffff);'
+}
+
+gen() {
+  TOKEN=$(generate_token)
+  mkdir -p ${TMP_DIR}
+  jsonnet -J ../../ --multi ${TMP_DIR} --tla-code-file cfg=${CONFIG_FILE} all.jsonnet
+  echo "kubeadm_token = \"$TOKEN\"" > ${CLUSTER_NAME}/terraform.tfvars
+}
+
+deploy() {
+  gen
+  terraform apply -var-file=${CLUSTER_NAME}/terraform.tfvars -state=${CLUSTER_NAME}/terraform.tfstate ${TMP_DIR}
+
+  PHASE2=$(jq -r '.phase2.provider' ${CONFIG_FILE})
+  if [[ "${PHASE2}" == "kubeadm" ]]; then
+    fetch_kubeconfig
+  fi
+}
+
+# For most implementations, the kubeconfig outputted by terraform is sufficient
+# to connect to the cluster. However, if using kubeadm, then "kubeadm init"
+# creates its own certificates on the master and ignores what was created by
+# terraform. In order to access the cluster remotely, we need to fetch the
+# proper kubeconfig from the master.
+fetch_kubeconfig() {
+  SSH_KEY_FILE=$(jq -r '.phase1.openstack.ssh_key_file' ${CONFIG_FILE})
+  MASTER=$(terraform output -json -state=${CLUSTER_NAME}/terraform.tfstate |jq -r ".[\"${CLUSTER_NAME}_master_ip\"].value")
+  SSHUSER=$(jq -r '.phase1.ssh_user' ${CONFIG_FILE})
+
+  if [[ -n "${SSHUSER}" ]];then
+    SSHUSER="${SSHUSER}@"
+  fi
+
+  for tries in {1..60}; do
+    echo Trying to fetch kubeconfig from master... $tries/60
+
+    # We can't use copy-files here because only root can read the file
+    # (intentionally). However, if the very first invocation of this command
+    # succeeds, then stdout is tainted by ssh initialization output, so we echo
+    # a marker to indicate the start of the file, and delete all lines up to and
+    # including it.
+    if ssh -i ${SSH_KEY_FILE} "$SSHUSER$MASTER" -o "StrictHostKeyChecking=no" -o "PasswordAuthentication=no" "echo STARTFILE; sudo cat /etc/kubernetes/admin.conf" > ${TMP_DIR}/kubeconfig.raw; then
+      sed '/^STARTFILE$/,$!d;/^STARTFILE$/d' ${TMP_DIR}/kubeconfig.raw > ${CLUSTER_NAME}/kubeconfig.json
```

Comment:
this places it in the following dir: `phase1/openstack/<clustername>/kubeconfig.json`. If we ever add more dirs to phase1/openstack, this could get buggy. Maybe we should add them to phase1/openstack/clusters/<clustername>

Reply:
Fixed this in a separate commit as i had take care of the change in gce too.